datapath: bigtcp: Fix the IPv4 BIG TCP may not work

haiyuewa opened the pull request:
The kernel will also update the IPv4 GRO/GSO setting if the new value of
"gso/gro_max_size" isn't greater than 65536:
  1. ip link set dev ens787np0 gso_max_size 130000
  2. ip link set dev ens787np0 gso_ipv4_max_size 128000
  3. ip -d -j link show dev ens787np0 | jq -c '.[0].gso_ipv4_max_size'
     128000
  4. ip link set dev ens787np0 gso_max_size 65536
  5. ip -d -j link show dev ens787np0 | jq -c '.[0].gso_ipv4_max_size'
     65536

Cilium will call IPv4 GRO/GSO setup firstly, then call IPv6. To handle the
above issue, setup the IPv6 GRO/GSO firstly, then IPv4 GRO/GSO.

PlatformLC commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -219,31 +219,31 @@ func InitBIGTCP(bigTCPConfig *Configuration) {
-		if haveIPv4 {
-			err = setGROGSOIPv4MaxSize(device,
-				bigTCPConfig.groIPv4MaxSize, bigTCPConfig.gsoIPv4MaxSize)
+		if haveIPv6 {
```
nit: should we leave an comment for this tricky order directly in code? Or hide such logic in setGROGSOIPv6MaxSize to make sure IPv6 operation will not impact IPv4 values.

haiyuewa replied:
This will make 'setGROGSOIPv6MaxSize' not so clean, by my understanding. And this logic is in kernel, git commit message should be fine, no bugs, then peope will be fine about this. ; - )

haiyuewa replied:
After thinking more, looks like make sense to add some comments. Thanks!

borkmann replied:
Thanks for the fix, @haiyuewa! Were you able to test it on ice with localy tweaking default{GSO,GRO}MaxSize? Would be curious what numbers you see before/after on ice.

Do you plan to follow-up with a knob for tweaking internal default{GSO,GRO}MaxSize?

haiyuewa replied:
I will re-submit with new logs after tweaking the MAX size.

And yes, if you don't start to code it. ;-) It should be about tweaking internal bigTCP{GRO/GSO}MaxSize, right ?

borkmann replied:
Back in IPv6 testing, @NikAleksandrov found the 196608 to be a sweet spot for mlx5.

I think for ice, we could read out link.Attrs().TSOMaxSize and probably determine this automatically even without knob.

For example, if the default{GSO,GRO}MaxSize fails to be set, then if link.Attrs().TSOMaxSize is still larger than 65536 we know that the device supports BIG TCP.

So if larger than 65536, we could take min(default{GSO,GRO}MaxSize, link.Attrs().TSOMaxSize) to figure it out.

borkmann replied:
@haiyuewa if you could help with testing on ice, I could code this up. But lets get first your fix from here in, so that we don't run into merge conflict.

haiyuewa replied:
@borkmann Looks like more things need to be considered. Sure, I can help on test it.

borkmann replied:
> Looks like more things need to be considered.

Could you elaborate?

haiyuewa replied:
> > Looks like more things need to be considered.
> 
> Could you elaborate?

I just thought it needs to do more things about "Back in IPv6 testing, @NikAleksandrov found the 196608 to be a sweet spot for mlx5." It's my misunderstanding.

Please let me know if the code is ready for ice test. Thanks.

borkmann replied:
Btw, I just checked the veth side, and here we should be good as v6 is adjusted before v4: https://github.com/cilium/cilium/pull/26172/commits/0ab201b39b55ab36e6bd6ba98d60f5da851afe73

PlatformLC commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -219,31 +219,31 @@ func InitBIGTCP(bigTCPConfig *Configuration) {
+				bigTCPConfig.gsoIPv4MaxSize, bigTCPConfig.groIPv4MaxSize)
 		}
 	}
 
```
nit: should we also modify the IPv4/IPv6 order for below err handling?

haiyuewa replied:
64KB will be OK. Just less code change to make things work.

borkmann commented on test/bigtcp/test.sh:
```diff
@@ -43,7 +43,7 @@ helm install cilium ${HELM_CHART_DIR} \
 
 # check if BIG TCP is initialized
-kubectl -n kube-system logs ds/cilium 2>&1 | grep "Setting up IPv6 BIG TCP"
+kubectl -n kube-system logs ds/cilium 2>&1 | grep "Setting up BIG TCP"
```
thanks for the fix. this could probably be submitted as a separate PR and we can already merge it.

haiyuewa replied:
Done: https://github.com/cilium/cilium/pull/26377

Commit:
datapath: bigtcp: Fix the IPv4 BIG TCP may not work

The kernel will also update the IPv4 GRO/GSO setting if the new value of
"gso/gro_max_size" isn't greater than 65536:

1. Enable IPv4 and IPv6 BIG TCP firstly:
  a). Dump the agent's log:
	$ kubectl -n kube-system logs cilium-tfhz5  2>&1 | grep "big-tcp"
	level=info msg="  --enable-ipv4-big-tcp='true'" subsys=daemon
	level=info msg="  --enable-ipv6-big-tcp='true'" subsys=daemon
	level=info msg="Setting up BIG TCP" subsys=big-tcp
	level=info msg="Setting IPv4 gso_max_size to 131072 and gro_max_size to 131072" device=enp0 subsys=big-tcp
	level=info msg="Setting IPv6 gso_max_size to 131072 and gro_max_size to 131072" device=enp0 subsys=big-tcp

  b). Check the GSO value on the host's net device:
	$ ip -d -j link show dev enp0 | jq -c '.[0].gso_max_size'
	131072

	$ ip -d -j link show dev enp0 | jq -c '.[0].gso_ipv4_max_size'
	131072

2. Then re-install the cilium by enabling IPv4 BIG TCP only:
  a). Dump the agent's log:
	$ kubectl -n kube-system logs cilium-zwpg6  2>&1 | grep "big-tcp"
	level=info msg="  --enable-ipv4-big-tcp='true'" subsys=daemon
	level=info msg="  --enable-ipv6-big-tcp='false'" subsys=daemon
	level=info msg="Setting up BIG TCP" subsys=big-tcp
	level=info msg="Setting IPv4 gso_max_size to 131072 and gro_max_size to 131072" device=enp0 subsys=big-tcp
	level=info msg="Setting IPv6 gso_max_size to 65536 and gro_max_size to 65536" device=enp0 subsys=big-tcp

  b). Check the GSO value on the host's net device:
	$ ip -d -j link show dev enp0 | jq -c '.[0].gso_max_size'
	65536

	$ip -d -j link show dev enp0 | jq -c '.[0].gso_ipv4_max_size'
	65536

No errors about the BIG TCP setting in cilium agent's log, but the value
of net device's '{gso,gro}_ipv4_max_size' is wrong.

So it needs to handle the IPv6 BIG TCP setting firstly, then IPv4.

Signed-off-by: Haiyue Wang <haiyue.wang@intel.com>
Link: https://lore.kernel.org/netdev/7e1f733cc96c7f7658fbf3276a90281b2f37acd1.1674921359.git.lucien.xin@gmail.com/

borkmann reviewed (approved):
Awesome, looks good to me! Thanks!
